plugins/flatpak/gs-flatpak.c

```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
 	return app;
 }
 
+gboolean
+gs_flatpak_app_get_copyable (GsFlatpak *self,
+			     GsApp *app,
+			     gboolean *copyable,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	if (copyable) {
+		*copyable = (gs_flatpak_get_collection_id (self, app) != NULL);
+	}
+
+	return TRUE;
+}
+
+typedef struct {
+	GsFlatpak *self;
+	GsApp *app;
+	GCancellable *cancellable;
+	gulong cancelled_id;
+	gboolean finished;
+	GError *error;
+} GsFlatpakCopyProcessHelper;
+
+static void
+gs_flatpak_copy_process_helper_free (GsFlatpakCopyProcessHelper *helper)
+{
+	g_object_unref (helper->app);
+	g_object_unref (helper->cancellable);
+	g_clear_error (&helper->error);
+	g_free (helper);
+}
+
+static GsFlatpakCopyProcessHelper *
+gs_flatpak_copy_process_helper_new (GsFlatpak *self,
+				    GsApp *app,
+				    GCancellable *cancellable,
+				    gulong cancelled_id)
+{
+	GsFlatpakCopyProcessHelper *helper = g_new0 (GsFlatpakCopyProcessHelper,
+						     1);
+	helper->self = self;
+	helper->app = g_object_ref (app);
+	helper->cancellable = g_object_ref (cancellable);
+	helper->cancelled_id = cancelled_id;
+	helper->finished = FALSE;
+	helper->error = NULL;
+	return helper;
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GsFlatpakCopyProcessHelper, gs_flatpak_copy_process_helper_free)
+
+static void
+copy_process_watch_cb (GPid pid, gint status, gpointer user_data)
+{
+	GsFlatpakCopyProcessHelper *helper =
+		(GsFlatpakCopyProcessHelper*) user_data;
+	g_autoptr(GError) error = NULL;
+
+	if (!g_cancellable_is_cancelled (helper->cancellable) && status) {
+		g_set_error (&helper->error,
+			     GS_PLUGIN_ERROR,
+			     GS_PLUGIN_ERROR_FAILED,
+			     "failed to copy flatpak app to removable "
+			     "media: `flatpak create-usb` failed with "
+			     "status %d", status);
+	}
+
+	g_cancellable_disconnect (helper->cancellable, helper->cancelled_id);
+	g_spawn_close_pid (pid);
+
+	helper->finished = TRUE;
+
+	/* once the copy terminates (successfully or not), set plugin status to
+	 * update UI accordingly */
+	gs_plugin_status_update (helper->self->plugin, helper->app,
+				 GS_PLUGIN_STATUS_FINISHED);
+
+	// update metadata to ensure the USB category includes this new app and
+	// we can immediately download an app, copy to USB, delete from
+	// computer, and install from USB without leaving the app page
+	gs_flatpak_refresh (helper->self, 0, GS_PLUGIN_REFRESH_FLAGS_METADATA,
```

Comment:
Are you sure this is the best way to do this? I think the refresh lists the apps in the repo which is quite slow, so maybe we should just modify the app as needed?